xdsl/irdl/declarative_assembly_format_parser.py

```diff
@@ -705,6 +706,39 @@ def parse_typeable_directive(self) -> TypeableDirective:
             return variable
         self.raise_error(f"unexpected token '{self._current_token.text}'")
 
+    def parse_custom_directive(self) -> FormatDirective:
+        """
+        Parse a custom directive, with the following format:
+          custom-directive ::= custom<bare-ident>((directive (`,` directive)*)?)
+        Assumes the keyword "custom" has already been parsed.
+        """
+
+        with self.in_angle_brackets():
+            name = self.parse_identifier()
+            if name not in self.op_def.custom_directives:
+                self.raise_error(f"Custom directive {name} cannot be found.")
+        directive = self.op_def.custom_directives[name]
+        if not issubclass(directive, FormatDirective):
+            self.raise_error(f"{name} is not a format directive.")
+        param_types = inspect.get_annotations(directive, eval_str=True)
```

Comment:
it would be great to do this once per directive definition and not once per op definition

Reply:
It evals once per field, which was quite visible on old traces of attribute definition, but I guess we can do this part later. The non-performance argument for it is that I feel like it's nice when the error occurs close in code to the thing that's broken, here if something goes wrong it's not at the point of definition, but at the point of use.

Reply:
The alternative here I guess is to have:
- Add a `CustomDirective` subclass with a classvar of `dict[str, type[FormatDirective]`
- a decorator `@irdl_custom_directive` to do the inspection which sets this.
- Use this instead of inspect in the above function.
?
I agree with the point about the better error handling. I think it should never happen that we have an unused custom directive, but it still seems nice that the directive itself would error instead of the operation that uses it. If the above sounds reasonable I can implement it.

Reply:
I think `@irdl_custom_directive` would be great, not sure what you mean with that classvar exactly but it sounds about right. It'll also make it easier to test I think.